File: telegram-bot-bumblebee-core/src/main/java/com/github/bumblebee/command/currency/CurrencyChartCommand.java
package com.github.bumblebee.command.currency;

import com.github.bumblebee.command.SingleArgumentCommand;
import com.github.bumblebee.command.currency.domain.DailyExchangeRate;
import com.github.bumblebee.command.currency.service.BYRExchangeRateChartService;
import com.github.bumblebee.command.currency.service.BYRExchangeRateStoreService;
import com.github.bumblebee.command.currency.service.ChartArgumentParser;
import com.github.bumblebee.service.RandomPhraseService;
import com.github.telegram.api.BotApi;
import com.github.telegram.domain.Update;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import org.xml.sax.SAXException;

import java.io.IOException;
import java.time.LocalDate;
import java.time.Period;
import java.util.List;

@Component
public class CurrencyChartCommand extends SingleArgumentCommand {

    private static final Logger log = LoggerFactory.getLogger(CurrencyChartCommand.class);

    private static final int BAR_TO_LINE_CHART_THRESHOLD_DAYS = 40;
    private static final int REDUCE_THRESHOLD_DAYS = 5 * 30;

    protected final BotApi botApi;
    protected final RandomPhraseService randomPhrase;
    protected final BYRExchangeRateStoreService exchangeRateStoreService;
    protected final BYRExchangeRateChartService chartService;
    protected final ChartArgumentParser argumentParser;

    @Autowired
    public CurrencyChartCommand(BotApi botApi,
                                RandomPhraseService randomPhrase,
                                BYRExchangeRateStoreService exchangeRateStoreService,
                                BYRExchangeRateChartService chartService,
                                ChartArgumentParser argumentParser) {
        this.botApi = botApi;
        this.randomPhrase = randomPhrase;
        this.exchangeRateStoreService = exchangeRateStoreService;
        this.chartService = chartService;
        this.argumentParser = argumentParser;
    }

    @Override
    public void handleCommand(Update update, long chatId, String argument) {

        ChartArgumentParser.DateRange range = argumentParser.getRange(argument);
        String errorMessage = validateRange(range.getFrom(), range.getTo());
        if (errorMessage != null) {
            botApi.sendMessage(chatId, errorMessage, update.getMessage().getMessageId());
            return;
        }
        List<String> currencies = argumentParser.getCurrencies(argument);

        buildChart(update, range.getFrom(), range.getTo(), currencies);
    }

    private String validateRange(LocalDate from, LocalDate to) {
        Period period = Period.between(from, to);
        if (period.getYears() > 3) {
            return "Period too large";
        }
        return null;
    }

    private void buildChart(Update update, LocalDate from, LocalDate to, List<String> currencies) {

        final Long chatId = update.getMessage().getChat().getId();

        Period period = Period.between(from, to);
        long approxDays = period.toTotalMonths() * 30 + period.getDays();
        boolean detailed = approxDays < BAR_TO_LINE_CHART_THRESHOLD_DAYS;
        boolean reduce = approxDays > REDUCE_THRESHOLD_DAYS;

        try {
            List<DailyExchangeRate> rates = exchangeRateStoreService.getRates(from, to, currencies);
            log.info("Fetched {} exchange rate records", rates.size());
            if (reduce) {
                rates = chartService.reduceToAverageByMonth(rates);
                log.info("{} records after reduction", rates.size());
            }

            byte[] png = chartService.createChartImage(rates, detailed, from, to);
            // todo: jpeg? WTF?
            botApi.sendPhoto(chatId, png, "image/jpeg", null, null, null, null);

        } catch (IOException | SAXException e) {
            log.error("Chart creation failed", e);
            botApi.sendMessage(chatId, randomPhrase.no(), update.getMessage().getMessageId());
        }
    }

}

File: telegram-bot-bumblebee-core/src/main/java/com/github/bumblebee/command/currency/CurrencyChartCommand.kt
package com.github.bumblebee.command.currency

import com.github.bumblebee.command.SingleArgumentCommand
import com.github.bumblebee.command.currency.service.BYRExchangeRateChartService
import com.github.bumblebee.command.currency.service.BYRExchangeRateStoreService
import com.github.bumblebee.command.currency.service.ChartArgumentParser
import com.github.bumblebee.service.RandomPhraseService
import com.github.bumblebee.util.logger
import com.github.telegram.api.BotApi
import com.github.telegram.api.InputFile
import com.github.telegram.domain.Update
import org.springframework.stereotype.Component
import org.xml.sax.SAXException
import java.io.ByteArrayInputStream
import java.io.IOException
import java.time.LocalDate
import java.time.Period

@Component
class CurrencyChartCommand(protected val botApi: BotApi,
                           protected val randomPhrase: RandomPhraseService,
                           protected val exchangeRateStoreService: BYRExchangeRateStoreService,
                           protected val chartService: BYRExchangeRateChartService,
                           protected val argumentParser: ChartArgumentParser) : SingleArgumentCommand() {

    override fun handleCommand(update: Update, chatId: Long, argument: String?) {
        val (from, to) = argumentParser.getRange(argument)
        validateRange(from, to)?.let { errorMessage ->
            botApi.sendMessage(chatId, errorMessage, update.message!!.messageId)
            return
        }
        val currencies = argumentParser.getCurrencies(argument)

        buildChart(update, from, to, currencies)
    }

    private fun validateRange(from: LocalDate, to: LocalDate): String? {
        val period = Period.between(from, to)
        return if (period.years > 3) {
            "Period too large"
        } else null
    }

    private fun buildChart(update: Update, from: LocalDate, to: LocalDate, currencies: List<String>) {
        val chatId = update.message!!.chat.id

        val period = Period.between(from, to)
        val approxDays = period.toTotalMonths() * 30 + period.days
        val detailed = approxDays < BAR_TO_LINE_CHART_THRESHOLD_DAYS
        val reduce = approxDays > REDUCE_THRESHOLD_DAYS

        try {
            var rates = exchangeRateStoreService.getRates(from, to, currencies)
            log.info("Fetched {} exchange rate records", rates.size)
            if (reduce) {
                rates = chartService.reduceToAverageByMonth(rates)
                log.info("{} records after reduction", rates.size)
            }

            val png = chartService.createChartImage(rates, detailed, from, to)
            botApi.sendPhoto(chatId, InputFile.photo(ByteArrayInputStream(png), "image"), null, null, null, null)
        } catch (e: IOException) {
            log.error("Chart creation failed", e)
            botApi.sendMessage(chatId, randomPhrase.no(), update.message!!.messageId)
        } catch (e: SAXException) {
            log.error("Chart creation failed", e)
            botApi.sendMessage(chatId, randomPhrase.no(), update.message!!.messageId)
        }
    }

    companion object {
        private val log = logger<CurrencyChartCommand>()

        private val BAR_TO_LINE_CHART_THRESHOLD_DAYS = 40
        private val REDUCE_THRESHOLD_DAYS = 5 * 30
    }

}
